### ISSUE TITLE
Has-one-field set to null becomes overwritten on loading lazy fields

### ISSUE THREAD

**Issue by sb-relaxt-at** [silverstripe/silverstripe-framework#9875]
## Affected Version

Version 4.7.2

## Description

This is a rather strange issue related to unsetting a has-one relation. I do have a HasManyList, where i remove some entries. This causes the ID fields to be set to null. When those removed objects are written by the HasManyList, LinkTracking kicks in, which in turn seems to perform loading of some lazy loaded fields. When lazy loading fields there is a check that collects all fields that need lazy loading, which includes all fields that are null. This results in the relation field being requeried and reset to the original state. Which in turns lead to the relation to be not removed properly.

## Steps to Reproduce

As I am not really aware of when lazy loading of fields occur I cannot properly reproduce this issue, I tried to fake it: 

````
$siteTree = new SiteTree();
$siteTree->ParentID = 1;
$siteTree->write();
// Fake existence of a lazy load field
$siteTree->setField('Sort_Lazy', SiteTree::class);
$siteTree->ParentID = null;
var_dump($siteTree->ParentID); // NULL
$siteTree->toMap(); // triggers loading of lazy fields
var_dump($siteTree->ParentID); // 1
````

**Comment by sminnee** [comment]
Interesting edge case!

While this is most likely a bug, I would note that a has_one ID field shouldn't ever be null in the database, and so we may want to case the null assignment to 0 in memory as part of fixing this.

I suspect that in the lazy-fetch logic there's an isset() check that should be an array_key_exists

Id also want to check this on a field that *can* store a null, such as (I think, from memory) a Varchar on a subclass.

**Comment by sminnee** [comment]
Marked as impact/low because it's an edge-case bug with a clear workaround.

FWIW I recommend setting has-one to 0 as the more accurate way of clearing it, in spite of this bug fix, as that's what the database will contain after clearing it.

**Comment by sb-relaxt-at** [comment]
I do agree, but there are at least two sections in the framework setting a has-one to null instead of 0, which caused the problem in the first place. (Sorry if I didn't made this clear): 
* [HasManyList#137](https://github.com/silverstripe/silverstripe-framework/blob/4/src/ORM/HasManyList.php#L137)
* [PolymorphicHasManyList#133](https://github.com/silverstripe/silverstripe-framework/blob/4/src/ORM/PolymorphicHasManyList.php#L133)

I suppose this to be the [problematic check in loadLazyField](https://github.com/silverstripe/silverstripe-framework/blob/4/src/ORM/DataObject.php#L2670).

**Comment by sminnee** [comment]
Yeah I think we should fix those internal set-to-nulls as well as fixing the underlying bug.

The fact that there's an explicit "or equals null" check suggests that someone (me?) really wanted a null value to be thought of as missing. It's probably worth looking at the git blame for that line to see what else was going on, and see if there any internal assignment-to-null operations related to lazy loading that should be replaced with isset.

**Comment by sminnee** [comment]
Thankfully, I think that changes a while back that give dataobject creation an explicit hydration step *probably* remove the change of any null values creeping in except from the database. But still worth a scan for "= null" to see what comes up.
